### ISSUE TITLE
Assets are compiled every time app starts

### ISSUE THREAD

**Issue by yivo** [rails/sprockets-rails#352]
After update from sprockets 2.x assets now are compiled every time app starts in development environment. What is exactly compiled? Everything! Application assets and assets from all loaded Rails::Engine.

To reproduce:
1. start app
2. open index page
3. wait until all assets are compiled
4. page load
5. if you reload page assets are not recompiled (as expected). 

Rails 4.2.5
sprockets 3.6.0
sprockets-rails 3.0.4

**Comment by rafaelfranca** [comment]
Correct. It is a know and desired behavior. It is compiled to check if all your assets used in assets helpers are in the precompile list. There is an option to opt out this behavior at the master branch if you want to disable this check.

**Comment by yivo** [comment]
Yes. It is good it can be disabled because for me first request takes about 70 seconds. 
Are speaking about [this](https://github.com/rails/sprockets-rails/blob/master/lib/sprockets/railtie.rb#L116)?
Installed `sprockets-rails` from master `gem 'sprockets-rails', github: 'rails/sprockets-rails', branch: 'master', require: 'sprockets/railtie'` and set `Rails.application.config.assets.check_precompiled_asset = true` but first request still compiles all assets.

**Comment by yivo** [comment]
Option `check_precompiled_asset ` will only prevent exception and doesn't affect compilation on first request.
https://github.com/rails/sprockets-rails/blob/b4f9006c7ff5ccd74b467c1b81e721c15dd5daf7/lib/sprockets/rails/helper.rb#L346

**Comment by rafaelfranca** [comment]
Yes, it does. The compilation is made when the `precompiled?` is called.

**Comment by yivo** [comment]
I made a mistake when set `check_precompiled_asset` to `true`. The correct value is `true`. Now all works as desired. Thank you.

**Comment by mipearson** [comment]
I would recommend adding something to the Rails upgrade guide regarding this - given the number of apps with significant asset compile times in the wild this could surprise a lot of people.

In our case, it caused significant test speed regressions in our parallel tests, as sixteen threads all tried to compile 50s worth of assets at the same time.

**Comment by tomca32** [comment]
While I like this behavior, it should definitely be mentioned in the upgrade guide.

**Comment by henrik** [comment]
We were bitten by this as well. I agree – a mention in the upgrade guide, or possibly outputting that it happens in the server output (e.g. "Checked precompiled assets (took 1234 seconds)") so it's easier to determine what's going on.

@rafaelfranca, would you be open to a PR with one or both of the above?

**Comment by rafaelfranca** [comment]
It doesn't fit the upgrade guide since you can upgrade rails without upgrading sprockets. This is a configuration in sprockets-rails and it is documented in sprockets-rails and it is in the CHANGELOG of sprockets-rails.

Maybe we could improve the sprockets-rails documentation, or maybe explain it in the assets pipeline guide?

**Comment by henrik** [comment]
I see your point. I'm not sure we would have found it in any of those places, but that's not to say it wouldn't help someone else.

How would you feel about something like "Checked precompiled assets (took 1234 seconds)" in the server output. But perhaps only showing it if takes longer than say 10 seconds, to keep down on noise for everyone else? I think that would have helped us catch it sooner.

**Comment by davidsbailey** [comment]
The performance benefits of disabling `config.assets.check_precompiled_asset` are not well documented, and are very difficult to discover without locating this github issue and reading the above discussion. @rafaelfranca , would you accept a pull request to document this more clearly? https://github.com/rails/sprockets-rails/pull/441

**Comment by prdetective** [comment]
Recently upgraded to sprockets 3.x. App has huge amount of assets - images and stylesheets. They should be loaded only under special conditions. With caching disabled, first request after server boot takes around ~1.5 minutes. 

I've spent couple of days trying to understand what's going on, until I finally found this answer: https://stackoverflow.com/questions/35683185/rails-sprocketsrailshelperassetnotprecompiled-in-development#51898048

Logging time-consuming operations would definetely have helped and would have saved a lot of time.
